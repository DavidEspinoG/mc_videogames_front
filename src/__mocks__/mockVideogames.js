export const mockVideogame = {
  id: 95,
  name: 'Street Fighter',
  photo: 'https://mms.businesswire.com/media/20230122005013/en/1692292/5/p1-1.jpg',
  description:
    'Street Fighter is a 1987 arcade game developed by Capcom. It is the first competitive fighting game produced by the company and the inaugural game in the Street Fighter series.',
  price_per_day: '20.0',
  created_at: '2023-07-20T21:47:16.294Z',
  updated_at: '2023-07-20T21:47:16.294Z',
};

const mockVideogames = [
  mockVideogame,
  {
    id: 94,
    name: 'Mortal Kombat',
    photo: 'https://thesource.com/wp-content/uploads/2017/10/862771.jpg',
    description:
      'Mortal Kombat is an American media franchise centered on a series of video games, originally developed by Midway Games in 1992.',
    price_per_day: '20.0',
    created_at: '2023-07-20T21:47:16.268Z',
    updated_at: '2023-07-20T21:47:16.268Z',
  },
];

export default mockVideogames;
