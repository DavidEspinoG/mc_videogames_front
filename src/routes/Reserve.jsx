import axios from 'axios';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useSearchParams } from 'react-router-dom';
import BASE_URL from '../redux/constants';
import { getVideogames } from '../redux/slices/videogamesSlice';
import '../styles/Login.scss';
import '../styles/reserve.scss';
import getTomorrowDate from '../utils/getTomorrowDate';

const Reserve = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const videogames = useSelector((state) => state.videogames.all);
  const jwt = useSelector((state) => state.user.jwt);
  const [searchParams] = useSearchParams();
  const videogameId = searchParams.get('videogameId');
  const [selectedVideogameId, setSelectedVideogameId] = useState(videogameId || 0);
  const [days, setDays] = useState(0);

  useEffect(() => {
    if (!videogames) {
      dispatch(getVideogames());
    }
  }, [dispatch, videogames]);

  const handleDateChange = (e) => {
    const now = new Date();
    const userDate = new Date(e.target.value);
    const difference = userDate.getTime() - now.getTime();
    const differenceDays = Math.ceil(difference / (1000 * 3600 * 24));
    setDays(differenceDays);
  };

  const fetchReservation = async () => {
    const url = `${BASE_URL}/reservations`;
    const body = { videogame_id: selectedVideogameId, days };
    const headers = {
      headers: {
        Authorization: jwt,
      },
    };
    const data = await axios.post(url, body, headers);
    return data;
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    await fetchReservation();
    navigate('/myReservations');
  };

  const handleVideogameChange = (e) => {
    setSelectedVideogameId(e.target.value);
  };

  return (
    <div className="reserve-main-container">
      <div className="container">
        <div className="login-form">
          <h3 className="subtitle text-center">RESERVE A VIDEOGAME</h3>
          <hr className="green-line" />
          <p>Select the videogame you want to rent and the date you want to return it</p>
          <form onSubmit={handleSubmit}>
            <div className="login-inputs">
              <select
                className="reserve-select"
                value={selectedVideogameId}
                onChange={handleVideogameChange}
                min={1}
              >
                <option value={0}>-- Select a videogame --</option>
                {videogames?.map((element) => (
                  <option key={element.id} value={element.id}>
                    {element.name}
                  </option>
                ))}
              </select>
              <input
                required
                type="date"
                min={getTomorrowDate()}
                className="input"
                onChange={handleDateChange}
              />
              <button type="submit" className="login-submit-button">
                Submit
              </button>
            </div>
          </form>
          {days ? (
            <p>
              * Your rent will last
              {' '}
              {days}
              {' '}
              days *
            </p>
          ) : ''}
        </div>
      </div>
    </div>
  );
};

export default Reserve;
