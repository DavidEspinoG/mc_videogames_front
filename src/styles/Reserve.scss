.reserve-select {
  appearance: none;
  color: white;
  font-family: var(--main-font);
  font-weight: 300;
  padding: 20px;
  border-radius: 30px;
  border: 1px solid white;
  background-color: var(--green);
  width: 270px;
}

.reserve-main-container {
  background-image: url('../assets/green_background.png');
  background-position: center;
  background-size: cover;
  
  height: 100vh;
  font-family: var(--main-font);
}


@media(min-width: 768px) {
  .reserve-main-container {
    width: calc(100vw - 240px);
  }
}
