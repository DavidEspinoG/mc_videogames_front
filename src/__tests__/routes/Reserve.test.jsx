import { vi } from 'vitest';
import Reserve from '../../routes/Reserve';
import mockVideogames from '../../__mocks__/mockVideogames';
import renderWithProviders from '../test-utils';

describe('Reserve', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should render as expected', () => {
    const date = new Date('2023/07/26');
    vi.setSystemTime(date);

    const { container } = renderWithProviders(<Reserve />, {
      preloadedState: {
        videogames: {
          all: mockVideogames,
        },
      },
    });
    expect(container).toMatchSnapshot();
  });
});
