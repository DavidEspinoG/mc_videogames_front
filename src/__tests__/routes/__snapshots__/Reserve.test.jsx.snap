// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reserve > should render as expected 1`] = `
<div>
  <div
    class="reserve-main-container"
  >
    <div
      class="container"
    >
      <div
        class="login-form"
      >
        <h3
          class="subtitle text-center"
        >
          RESERVE A VIDEOGAME
        </h3>
        <hr
          class="green-line"
        />
        <p>
          Select the videogame you want to rent and the date you want to return it
        </p>
        <form>
          <div
            class="login-inputs"
          >
            <select
              class="reserve-select"
              min="1"
            >
              <option
                value="0"
              >
                -- Select a videogame --
              </option>
              <option
                value="95"
              >
                Street Fighter
              </option>
              <option
                value="94"
              >
                Mortal Kombat
              </option>
            </select>
            <input
              class="input"
              min="2023-07-27"
              required=""
              type="date"
            />
            <button
              class="login-submit-button"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
